2021-09-30-review-apps-github-labels.md

```diff
@@ -0,0 +1,99 @@
+
+The workflows configuration will look something like this:
+
+```
+name: Review Apps
+on:
+  pull_request:
+    types: [labeled]
+
+jobs:
+  deploy-review-app:
+    if: ${{ github.event.label.name == 'deploy-review-app' }}
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: Deploy Review App
+        run: echo "Deploy!"
+
+  destroy-review-app:
+    if: ${{ github.event.label.name == 'destroy-review-app' }}
+    runs-on: ubuntu-latest
+
+    steps:
+      - uses: actions/checkout@v1
+      - name: Destroy Review App
+        run: echo "Destroy!"
+```
+
+for each of those jobs we would actually use a GitHub Action instead of just echoing some text.
+
+### GitHub Actions:
+
+We would need to implement 2 public GitHub Actions (one for creating/updating review apps, and one for destroying them), but the logic will be similar in both cases so we'll describe it in a generic way.
+
+The GitHub Actions will have this logic:
+
+- Include heroku's and github's npm modules to handle the different API requests that we'll need
+- Check that the user doing the action is an admin/maintainer/contributor of the repo so random people won't be able to trigger Review Apps creation adding labels (so this is safe to use in OSS projects)
+- Get all the info needed from the PR and the repo's secret (for security) to generate the payload for Heroku's API
+- Trigger the Review App concrete action (create/update/destroy)
+- Remove the label that triggered this action using Github's API
+
+### Resources
+
+We've found some resources and existing GitHub Actions to use as reference for most of the code needed:
+
+- Create Heroku's review app https://github.com/mheap/github-action-pr-heroku-review-app (this includes using the npm module, checking the user permissions, preparing the payload, hitting Heroku's `create` endpoint for review apps)
+- Add labels to a PR https://github.com/TimonVS/pr-labeler-action (this includes connecting to the GitHub API to set labels for a PR, we can use it as reference for removing the label instead)
+- How to create Github Actions https://docs.github.com/en/actions/creating-actions/creating-a-javascript-action (this shows the step-by-step on creating github actions)
+- Documentation on setting ENV variables for the GitHub Actions https://docs.github.com/en/actions/learn-github-actions/environment-variables (needed for the Heroku's API token and Pipeline ID)
+
+## Complete workflow:
+
+- A developer creates a PR (no review app is created)
+- When the PR is ready for QA and the person doing the QA is ready, they will add the `deploy-review-app` label to the PR
+- After a moment, the GitHub Action will be triggered, creating a Review App and removing the `deploy-review-app` label (the information of the deploy should be added to the PR by Heroku like it's currently doing)
+- When the QA is done, the QA person will add the `destroy-review-app` label to the PR
```

Comment:
I like how this sounds and I think it makes sense to start this way. But I also see a lot of room for human error here. We've had problems before with people forgetting the Ready to Merge label, so there's reason to believe they'd forget this one as well.

Might be worth exploring for a future iteration here if we can integrate Jira into this mix. And then once the story goes on to `Done` the review app can be destroyed regardless of label?